File: app/routers/tools.py
from __future__ import annotations

from fastapi import APIRouter, Form, Request
from fastapi.responses import RedirectResponse

from app.deps import CurrentUser, CurrentWorkspace, DbSession, flash, get_site_or_404
from app.models import JobType, PageUrl
from app.services import indexer
from app.services.seo_tools import fetch_page_meta
from app.services.sitemap_parser import crawl_sitemap
from app.services.sites import list_sites
from app.services.urls import normalize_url, parse_url_blob
from app.templating import render

router = APIRouter(prefix="/tools", tags=["tools"])


@router.get("")
def tools_page(request: Request, user: CurrentUser, workspace: CurrentWorkspace, db: DbSession):
    return render(
        request,
        "tools.html",
        {
            "user": user,
            "workspace": workspace,
            "sites": list_sites(db, workspace),
            "active_page": "tools",
            "tool": request.query_params.get("tool", "instant"),
        },
    )


@router.post("/instant")
def instant_index(
    request: Request,
    user: CurrentUser,
    workspace: CurrentWorkspace,
    db: DbSession,
    site_id: int = Form(...),
    urls_blob: str = Form(...),
    notification_type: str = Form("URL_UPDATED"),
):
    site = get_site_or_404(db, workspace, site_id)
    candidates = parse_url_blob(urls_blob)
    if not candidates:
        flash(request, "Nie podano poprawnych adresow URL.", "error")
        return RedirectResponse("/tools?tool=instant", status_code=303)

    job_type = JobType.URL_DELETED if notification_type == "URL_DELETED" else JobType.URL_UPDATED
    results = {"success": 0, "failed": 0, "messages": []}

    from sqlalchemy import select

    for url in candidates[:50]:
        page = db.scalar(select(PageUrl).where(PageUrl.site_id == site.id, PageUrl.url == url))
        if page is None:
            page = PageUrl(site_id=site.id, url=url, source="manual")
            db.add(page)
            db.flush()

        job = indexer.submit_single(
            db, user, workspace, site, page, url, job_type=job_type, triggered_by="manual"
        )
        if job.status == "SUCCESS":
            results["success"] += 1
        else:
            results["failed"] += 1
            if job.message:
                results["messages"].append(f"{url}: {job.message}")

    if results["success"]:
        flash(request, f"Zgloszono {results['success']} URL-i do Google.", "success")
    for message in results["messages"][:3]:
        flash(request, message, "error")

    return RedirectResponse("/tools?tool=instant", status_code=303)


@router.post("/indexnow")
def indexnow_submit(
    request: Request,
    workspace: CurrentWorkspace,
    db: DbSession,
    site_id: int = Form(...),
    urls_blob: str = Form(...),
):
    site = get_site_or_404(db, workspace, site_id)
    candidates = parse_url_blob(urls_blob)
    if not candidates:
        flash(request, "Nie podano poprawnych adresow URL.", "error")
        return RedirectResponse("/tools?tool=indexnow", status_code=303)

    result = indexer.submit_indexnow(db, site, candidates)
    flash(
        request,
        f"IndexNow: {result.get('message')}",
        "success" if result.get("ok") else "error",
    )
    return RedirectResponse("/tools?tool=indexnow", status_code=303)


@router.get("/preview")
def preview_tool(
    request: Request, user: CurrentUser, workspace: CurrentWorkspace, db: DbSession, url: str = ""
):
    meta = None
    if url:
        normalized = normalize_url(url)
        meta = fetch_page_meta(normalized) if normalized else {"error": "Nieprawidlowy URL."}

    return render(
        request,
        "tools.html",
        {
            "user": user,
            "workspace": workspace,
            "sites": list_sites(db, workspace),
            "meta": meta,
            "preview_url": url,
            "tool": "preview",
            "active_page": "tools",
        },
    )


@router.get("/sitemap")
def sitemap_tool(
    request: Request, user: CurrentUser, workspace: CurrentWorkspace, db: DbSession, url: str = ""
):
    result = None
    if url:
        normalized = normalize_url(url)
        if normalized:
            crawled = crawl_sitemap(normalized)
            result = {
                "source": crawled.source,
                "is_index": crawled.is_index,
                "error": crawled.error,
                "child_sitemaps": crawled.child_sitemaps,
                "count": len(crawled.entries),
                "entries": crawled.entries[:500],
            }
        else:
            result = {"error": "Nieprawidlowy URL."}

    return render(
        request,
        "tools.html",
        {
            "user": user,
            "workspace": workspace,
            "sites": list_sites(db, workspace),
            "sitemap_result": result,
            "sitemap_url": url,
            "tool": "sitemap",
            "active_page": "tools",
        },
    )

File: src/routes/tools.ts
import { Router } from "express";
import { db } from "../db/index.js";
import { JobStatus, JobType, type PageUrl } from "../db/types.js";
import { sha256 } from "../lib/crypto.js";
import { asyncHandler, flash, requireSite } from "../middleware/auth.js";
import * as indexer from "../services/indexer.js";
import { fetchPageMeta } from "../services/seoTools.js";
import { crawlSitemap } from "../services/sitemapParser.js";
import { listSites } from "../services/sites.js";
import { normalizeUrl, parseUrlBlob } from "../services/urls.js";
import { baseContext } from "../templating.js";
import { panelAuth } from "./auth.js";

export const toolsRouter = Router();
toolsRouter.use(...panelAuth);

toolsRouter.get(
  "/",
  asyncHandler(async (req, res) => {
    res.render(
      "tools.html",
      baseContext(req, {
        user: req.user,
        workspace: req.workspace,
        sites: await listSites(req.workspace!.id),
        active_page: "tools",
        tool: typeof req.query.tool === "string" ? req.query.tool : "instant",
      }),
    );
  }),
);

toolsRouter.post(
  "/instant",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.body.site_id));
    const candidates = parseUrlBlob(String(req.body.urls_blob ?? ""));
    if (candidates.length === 0) {
      flash(req, "Nie podano poprawnych adresow URL.", "error");
      return res.redirect(303, "/tools?tool=instant");
    }

    const jobType =
      req.body.notification_type === "URL_DELETED" ? JobType.URL_DELETED : JobType.URL_UPDATED;
    const results = { success: 0, failed: 0, messages: [] as string[] };
    const ctx = {
      userId: req.user!.id,
      userEmail: req.user!.email,
      workspace: req.workspace!,
    };

    for (const url of candidates.slice(0, 50)) {
      let page: PageUrl | undefined = await db
        .selectFrom("urls")
        .selectAll()
        .where("site_id", "=", site.id)
        .where("url_hash", "=", sha256(url))
        .executeTakeFirst();

      if (!page) {
        const inserted = await db
          .insertInto("urls")
          .values({
            site_id: site.id,
            url: url.slice(0, 2048),
            url_hash: sha256(url),
            source: "manual",
          })
          .executeTakeFirst();
        page = await db
          .selectFrom("urls")
          .selectAll()
          .where("id", "=", Number(inserted.insertId))
          .executeTakeFirstOrThrow();
      }

      const job = await indexer.submitSingle(ctx, site, page, url, jobType, "manual");
      if (job.status === JobStatus.SUCCESS) results.success += 1;
      else {
        results.failed += 1;
        if (job.message) results.messages.push(`${url}: ${job.message}`);
      }
    }

    if (results.success) {
      flash(req, `Zgloszono ${results.success} URL-i do Google.`, "success");
    }
    for (const message of results.messages.slice(0, 3)) flash(req, message, "error");
    res.redirect(303, "/tools?tool=instant");
  }),
);

toolsRouter.post(
  "/indexnow",
  asyncHandler(async (req, res) => {
    const site = await requireSite(req, Number(req.body.site_id));
    const candidates = parseUrlBlob(String(req.body.urls_blob ?? ""));
    if (candidates.length === 0) {
      flash(req, "Nie podano poprawnych adresow URL.", "error");
      return res.redirect(303, "/tools?tool=indexnow");
    }
    const result = await indexer.submitIndexNow(site, candidates);
    flash(req, `IndexNow: ${result.message}`, result.ok ? "success" : "error");
    res.redirect(303, "/tools?tool=indexnow");
  }),
);

toolsRouter.get(
  "/preview",
  asyncHandler(async (req, res) => {
    const url = typeof req.query.url === "string" ? req.query.url : "";
    let meta = null;
    if (url) {
      const normalized = normalizeUrl(url);
      meta = normalized
        ? await fetchPageMeta(normalized)
        : { url, error: "Nieprawidlowy URL." };
    }
    res.render(
      "tools.html",
      baseContext(req, {
        user: req.user,
        workspace: req.workspace,
        sites: await listSites(req.workspace!.id),
        meta,
        preview_url: url,
        tool: "preview",
        active_page: "tools",
      }),
    );
  }),
);

toolsRouter.get(
  "/sitemap",
  asyncHandler(async (req, res) => {
    const url = typeof req.query.url === "string" ? req.query.url : "";
    let result = null;
    if (url) {
      const normalized = normalizeUrl(url);
      if (normalized) {
        const crawled = await crawlSitemap(normalized);
        result = {
          source: crawled.source,
          is_index: crawled.isIndex,
          error: crawled.error,
          child_sitemaps: crawled.childSitemaps,
          count: crawled.entries.length,
          entries: crawled.entries.slice(0, 500),
        };
      } else {
        result = { error: "Nieprawidlowy URL." };
      }
    }
    res.render(
      "tools.html",
      baseContext(req, {
        user: req.user,
        workspace: req.workspace,
        sites: await listSites(req.workspace!.id),
        sitemap_result: result,
        sitemap_url: url,
        tool: "sitemap",
        active_page: "tools",
      }),
    );
  }),
);
